src/wallet/api/wallet2_api.h

```diff
@@ -1144,6 +1154,519 @@ struct Wallet
+    * brief: useForkRules - check if the wallet uses the rules for hard fork version at current block height - early_blocks
+    * param: version - hard fork version
+    * param: early_blocks -
+    * return: true if fork rules are used
+    * note: sets status error on fail
+    */
+//    virtual bool useForkRules(std::uint8_t version, std::int64_t early_blocks) const = 0;
+// TODO : wallet2::transfer_details
+    /**
+    * brief: getTransferDetails -
+    * param: idx - index in transfer storage
+    * return: transfer details for given index
+    * note: sets status error on fail
+    */
+//    virtual const wallet2::transfer_details &getTransferDetails(std::size_t idx) const = 0;
+    /**
+    * brief: discardUnmixableOutputs - freeze all unmixable outputs
+    * note: sets status error on fail
+    */
+    virtual void discardUnmixableOutputs() = 0;
+    /**
+    * brief: setTxKey - set the transaction key (r) for a given <txid> in case the tx was made by some other device or 3rd party wallet
+    * param: txid -
+    * param: tx_key - secret transaction key r
+    * param: additional_tx_keys -
+    * param: single_destination_subaddress -
+    * note: sets status error on fail
+    */
+    // QUESTION : There were some cases already where I wished to have boost::optional in here. I have no explicit example, but in general e.g. for functions that return a bool, but have to return early on error, it would be cleaner to return boost::none instead of true/false, because the API user could forget to check the error status and may use an incorrect value. Similar for uint*_t, where we can't use e.g. -1 on error.
+    //            If we add access to boost::optional revisit all new functions to see which would benefit from it.
+    virtual void setTxKey(const std::string &txid, const std::string &tx_key, const std::vector<std::string> &additional_tx_keys, const boost::optional<std::string> &single_destination_subaddress) = 0;
+    /**
+    * brief: getDaemonAddress -
+    * return: daemon address
+    */
+    virtual std::string getDaemonAddress() const = 0;
+    /**
+    * brief: getDaemonAdjustedTime -
+    * return: daemon adjusted time
+    * note: sets status error on fail
+    */
+    virtual std::uint64_t getDaemonAdjustedTime() const = 0;
+    /**
+    * brief: setCacheDescription - set wallet cache attribute with key ATTRIBUTE_DESCRIPTION
+    * param: description -
+    */
+    virtual void setCacheDescription(const std::string &description) = 0;
+    /**
+    * brief: getCacheDescription - get wallet cache attribute with key ATTRIBUTE_DESCRIPTION
+    * return: description
+    */
+    virtual std::string getCacheDescription() const = 0;
+    /**
+    * brief: getAccountTags - get the list of registered account tags
+    * return: first.Key=(tag's name), first.Value=(tag's label), second[i]=(i-th account's tag)
+    */
+    virtual const std::pair<std::map<std::string, std::string>, std::vector<std::string>>& getAccountTags() = 0;
```

Comment:
Seems to me, if we "go with the pattern", we have a method somewhere that gives back all existing accounts with info, using an `account_info` struct. The tag would be part of that struct, and thus this call here not needed because not "essential".

Or maybe a call giving back all subaddresses, with info about them, and info how they group into accounts - and any tags of course.

Methods like the following `setAccountTag` stay of course even if we use this pattern whenever we can, because changing things is something fundamentally different and altogether impossible if there is no method, after all.

Reply:
I think that _tags_ would fit into `SubaddressAccountRow`, but returning the "i-th account's tag" is just one part of this function, it also returns all tag descriptions (used by [wallet-cli](https://github.com/monero-project/monero/blob/a1dc85c5373a30f14aaf7dcfdd95f5a7375d3623/src/simplewallet/simplewallet.cpp#L9360-L9374) and [wallet-rpc](https://github.com/monero-project/monero/blob/a1dc85c5373a30f14aaf7dcfdd95f5a7375d3623/src/wallet/wallet_rpc_server.cpp#L767-L785)). So we probably need at least a `getAccountTagDescriptions()` method anyways, or alternatively we could store the tag description in `SubaddressAccountRow`, too.